src/autosave.js

```diff
@@ -0,0 +1,288 @@
+		 * the model's data changes. It might be called some time after the change,
+		 * since the event is throttled for performance reasons.
+		 *
+		 * @type {module:autosave/autosave~SaveProvider}
+		 */
+		this.provider = undefined;
+
+		/**
+		 * Throttled save method.
+		 *
+		 * @protected
+		 * @type {Function}
+		 */
+		this._throttledSave = throttle( this._save.bind( this ), 500 );
+
+		/**
+		 * Last document version.
+		 *
+		 * @protected
+		 * @type {Number}
+		 */
+		this._lastDocumentVersion = editor.model.document.version;
+
+		/**
+		 * @private
+		 * @type {DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DomEmitterMixin );
+
+		/**
+		 * Save action counter monitors number of actions.
+		 *
+		 * @private
+		 * @type {Number}
+		 */
+		this._saveActionCounter = 0;
+
+		/**
+		 * An action that will be added to pending action manager for actions happening in that plugin.
+		 *
+		 * @private
+		 * @type {Object|null}
+		 */
+		this._action = null;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const doc = editor.model.document;
+		const pendingActions = editor.plugins.get( PendingActions );
+
+		this.listenTo( doc, 'change:data', () => {
+			this._incrementCounter();
+
+			const willOriginalFunctionBeCalled = this._throttledSave();
+
+			if ( !willOriginalFunctionBeCalled ) {
+				this._decrementCounter();
+			}
+		} );
+
+		// Flush on the editor's destroy listener with the highest priority to ensure that
+		// `editor.getData()` will be called before plugins are destroyed.
+		this.listenTo( editor, 'destroy', () => this._flush(), { priority: 'highest' } );
+
+		// It's not possible to easy test it because karma uses `beforeunload` event
+		// to warn before full page reload and this event cannot be dispatched manually.
+		/* istanbul ignore next */
+		this._domEmitter.listenTo( window, 'beforeunload', ( evtInfo, domEvt ) => {
+			if ( pendingActions.isPending ) {
+				domEvt.returnValue = pendingActions.first.message;
+			}
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	destroy() {
+		// There's no need for canceling or flushing the throttled save, as
+		// it's done on the editor's destroy event with the highest priority.
+
+		this._domEmitter.stopListening();
+		super.destroy();
+	}
+
+	/**
+	 * Invokes remaining `_save` method call.
+	 *
+	 * @protected
+	 */
+	_flush() {
+		this._throttledSave.flush();
+	}
+
+	/**
+	 * If the provider is set and new document version exists,
+	 * `_save()` method creates a pending action and calls `provider.save()` method.
+	 * It waits for the result and then removes the created pending action.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this.editor.model.document.version;
+
+		// Marker's change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( !this.provider || version < this._lastDocumentVersion ) {
```

Comment:
I agree, but we'd have to assume that the editor's data is a string. For now, it might not be a string.

https://github.com/ckeditor/ckeditor5-core/blob/c31d691db86cfa37d1bc37d0f56fd812fca1b9be/src/editor/utils/dataapimixin.js#L34-L43

Reply:
I see there:
```
 * @method #getData
 * @returns {String} Output data.
```
So, I don't see a reason how data may not be a string.

Reply:
That comes from a mixin and it's not a method of the base `Editor` class. But we even don't have such editor without the `getData()` method at this moment.